### ISSUE TITLE
Circular dependency with jaraco.text

### ISSUE THREAD

**Issue by jaraco** [jaraco/jaraco.collections#4]
As reported in jaraco/jaraco.functools#9, this project has a circular dependency with `jaraco.text`.

@fridh asks:

> Was this intentional? We can build wheels separately for them (as we currently do), however, it requires us to install them together, which is a bit unfortunate, as we typically have a separate `site-packages` folder for each package (cc @dotlambda).

**Comment by jaraco** [comment]
The circular dependency is intentional, although only incidentally - as they depend on functionality from each other. The circular dependency is only a runtime dependency, meaning it should be possible to build each one independently and install them together.  I'm not sure what you mean by "it requires us to install them together". Is there a reason they can't be installed separately like any other package and its dependencies? Here I install jaraco.collections to one path and jaraco.text to another and they play nicely together, and I'd expect the same for other installers/environments.

```
work $ pip install -t path1 --no-deps jaraco.collections                                                                                                                                   
Collecting jaraco.collections
  Using cached https://files.pythonhosted.org/packages/73/22/c81ffbf17b7c2ab650f363cd9cedb6a950b4b27a17ef806febb2cd2026d9/jaraco.collections-2.0-py2.py3-none-any.whl
Installing collected packages: jaraco.collections
Successfully installed jaraco.collections-2.0
work $ pip install -t path2 --no-deps jaraco.text                                                                                                                                          
Collecting jaraco.text
  Using cached https://files.pythonhosted.org/packages/74/f2/40a800c2f9bc1a53393754acb6e897a6142fc271539e22efd429db21f3ff/jaraco.text-2.0-py2.py3-none-any.whl
Installing collected packages: jaraco.text
Successfully installed jaraco.text-2.0
work $ env PYTHONPATH=path1:path2 python3                                                                                                                                                  
Python 3.7.2 (v3.7.2:9a3ffc0492, Dec 24 2018, 02:44:43) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import jaraco.collections
>>> import jaraco.text
>>> jaraco.collections.__file__
'/Users/jaraco/draft/work/path1/jaraco/collections.py'
>>> jaraco.text.__file__
'/Users/jaraco/draft/work/path2/jaraco/text.py'
```

**Comment by FRidh** [comment]
We can install them indeed using `--no-deps`. It's just from a user point of view that the individual package would then be "broken" because a dependency is missing. Normally we declare dependencies and they get automatically added, however, our dependency graph has to be acyclic.

As I wrote, we can work around it, but it's a bit ugly.

**Comment by jaraco** [comment]
I agree it would be nice for all packages to be acyclic. But I don't think there's a standard anywhere that guarantees such a graph. I could address the situation by vendoring some of the functionality and violating DRY and separation of concerns. Or I could combine both of these libraries, but then what would I call it? `jaraco.collections-and-text` is a bit unwieldy. 

I suppose I could revive that concept and create a `jaraco.std` or similar that collects all of these smaller packages together again.These packages came out of a "kitchen sink" project I had called [jaraco.util](/jaraco/jaraco.util) (a [concept bemoaned here](http://www.adam-bien.com/roller/abien/entry/util_packages_are_evil))... and others complained that the project seemed to change too often for concerns unrelated to the usage for which it was employed.

I've looked at the code, and the DictStack usage in jaraco.text could be replaced with an inline copy... and in fact, that function is barely used and obviated by f-strings and [future-fstrings](https://github.com/asottile/future-fstrings), so maybe I just remove that functionality altogether.

Still, I don't think that solves the problem in general. If any project in any dependency tree should wish to mutually depend on functionality from the other, what should that project do? Should the packaging tools warn about this (I know pip-compile does)? I'd feel much better following a generally-accepted and recommended approach.

**Comment by jaraco** [comment]
I'm going to reserve this issue for the specific issue between collections and text in jaraco. If you wish to pursue the issue more generally, I recommend doing so in [packaging-problems](/pypa/packaging-problems). Until there's an accepted standard practice, this issue may re-emerge.
